### PR TITLE
Trackers rewrite & Switch to new math lib

### PR DESCRIPTION
**Pull request by Louka3000**
- Rewrites the trackers to a single class with flags and logic handlers instead of abstractions.
- Use a new immutable 3d math lib by AxisAngle.

Improvements (use for 0.7.2 changelog):
- Makes default tracker settings consistent (individual drift comp. always enabled)
- Improve drift compensation a lot! (Works lying down)
- Disconnected or errored trackers won’t affect skeleton until they go back to OK

### CODE REVIEW

**Comment by ImUrX** [comment]
this was suffering

**Comment by Louka3000** [comment]
Note: BVH is broken. Currently in public beta test: https://discord.com/channels/817184208525983775/1093712010740502578

File: server/src/main/java/dev/slimevr/tracking/processor/BoneInfo.java [hunk 1]
```diff
@@ -13,7 +16,12 @@ public class BoneInfo {
 	public final TransformNode headNode;
 	public final TransformNode tailNode;
 	public float length;
-	private static final Quaternion FOOT_OFFSET = Quaternion.X_90_DEG;
+	private static final Quaternion FOOT_OFFSET = new EulerAngles(
+		EulerOrder.YZX,
+		FastMath.HALF_PI,
+		0f,
+		0f
+	).toQuaternion();
```
Comment (ButterscotchV) [inline server/src/main/java/dev/slimevr/tracking/processor/BoneInfo.java:24]: I'm sure this can be done without euler angles?

File: server/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweakBuffer.java [hunk 2]
```diff
@@ -518,38 +462,50 @@ private int checkStateRight() {
 	// get the difference in feet position between the kinematic and corrected
 	// positions of the feet disregarding vertical displacement
 	private float getLeftFootHorizontalDifference() {
-		return leftFootPositionCorrected.subtract(leftFootPosition).setY(0).length();
+		return new Vector3(
+			leftFootPositionCorrected.minus(leftFootPosition).getX(),
+			0f,
+			leftFootPositionCorrected.minus(leftFootPosition).getZ()
+		).len();
```
Comment (ButterscotchV) [inline server/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweakBuffer.java:469]: Why are you calculating this twice...?
```suggestion
		Vector3 diff = leftFootPositionCorrected.minus(leftFootPosition);
		return new Vector3(
			diff.getX(),
			0f,
			diff.getZ()
		).len();
```

File: server/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweakBuffer.java [hunk 3]
```diff
@@ -518,38 +462,50 @@ private int checkStateRight() {
...
 
 	// get the difference in feet position between the kinematic and corrected
 	// positions of the feet
 	private float getRightFootHorizontalDifference() {
-		return rightFootPositionCorrected.subtract(rightFootPosition).setY(0).length();
+		return new Vector3(
+			rightFootPositionCorrected.minus(rightFootPosition).getX(),
+			0f,
+			rightFootPositionCorrected.minus(rightFootPosition).getZ()
+		).len();
```
Comment (ButterscotchV) [inline server/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweakBuffer.java:479]: Same as last comment, why are you calculating this twice?
```suggestion
		Vector3 diff = rightFootPositionCorrected.minus(rightFootPosition);
		return new Vector3(
			diff.getX(),
			0f,
			diff.getZ()
		).len();
```

File: server/src/main/java/dev/slimevr/tracking/processor/skeleton/ViveEmulation.java [hunk 4]
```diff
@@ -84,61 +84,65 @@ public void update() {
...
-			skeleton.computedWaistTracker.position.set(getFlyingPos());
-			skeleton.computedWaistTracker.rotation.set(getFlyingRotation());
+			skeleton.computedHipTracker.setPosition(getFlyingPos());
+			skeleton.computedHipTracker.setRotation(getFlyingRotation());
 		}
 		// in this state the tracker is returning to its position but has not
 		// yet overshot its target
 		else if (flyingBack && overShooting) {
-			if (targetPosition.distanceSquared(lastPosition) < NEARLY_ZERO * NEARLY_ZERO) {
+			if (targetPosition.minus(lastPosition).lenSq() < NEARLY_ZERO * NEARLY_ZERO) {
```
Comment (ButterscotchV) [inline server/src/main/java/dev/slimevr/tracking/processor/skeleton/ViveEmulation.java:96]: This is mathematically fine, but I want `b - a` order for clarity.
```suggestion
			if (lastPosition.minus(targetPosition).lenSq() < NEARLY_ZERO * NEARLY_ZERO) {
```

File: server/src/main/java/dev/slimevr/tracking/processor/skeleton/ViveEmulation.java [hunk 5]
```diff
@@ -84,61 +84,65 @@ public void update() {
...
 				overShooting = false;
 				flySpeed = FLY_BACK_OVERSHOOT;
-				targetPosition.set(skeleton.computedWaistTracker.position);
+				targetPosition = skeleton.computedHipTracker.getPosition();
 				return;
 			}
-			skeleton.computedWaistTracker.position.set(getFlyingBackPos());
+			skeleton.computedHipTracker.setPosition((getFlyingBackPos()));
```
Comment (ButterscotchV) [inline server/src/main/java/dev/slimevr/tracking/processor/skeleton/ViveEmulation.java:102]: Unecessary brackets.
```suggestion
			skeleton.computedHipTracker.setPosition(getFlyingBackPos());
```

File: server/src/main/java/dev/slimevr/tracking/processor/skeleton/ViveEmulation.java [hunk 6]
```diff
@@ -84,61 +84,65 @@ public void update() {
...
 		}
 		// in this state the tracker will return to its original position
 		else if (flyingBack) {
-			if (skeleton.computedWaistTracker.position.distance(lastPosition) < NEARLY_ZERO) {
+			if (
+				skeleton.computedHipTracker.getPosition().minus(lastPosition).len() < NEARLY_ZERO
```
Comment (ButterscotchV) [inline server/src/main/java/dev/slimevr/tracking/processor/skeleton/ViveEmulation.java:107]: This is mathematically fine, but I want `b - a` order for clarity.
```suggestion
				lastPosition.minus(skeleton.computedHipTracker.getPosition()).len() < NEARLY_ZERO
```

File: server/src/main/java/dev/slimevr/tracking/trackers/TrackerPosition.java [hunk 7]
```diff
@@ -1,148 +0,0 @@
...
-		} else if (designation.equalsIgnoreCase("body:right_upperarm")) {
-			designation = "body:right_upper_arm";
-		} else if (designation.equalsIgnoreCase("body:left_controller")) {
-			designation = "body:left_hand";
-		} else if (designation.equalsIgnoreCase("body:right_controller")) {
-			designation = "body:right_hand";
-		}
-
-		return Optional.ofNullable(byDesignation.get(designation.toLowerCase()));
-	}
```
Comment (ButterscotchV) [inline server/src/main/java/dev/slimevr/tracking/trackers/TrackerPosition.java:135]: Doesn't removing this break some backwards compatability? I don't think this should be removed...

File: server/src/main/java/io/github/axisangles/ktmath/Transform.kt [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+	var rotation: Quaternion = Quaternion.IDENTITY,
+	var translation: Vector3 = Vector3.NULL,
+	var scale: Vector3 = Vector3(1f, 1f, 1f),
+) {
+	companion object {
+		val IDENTITY = Transform()
+	}
+
+	fun set(transform: Transform): Transform {
+		// FIXME: Should those attributes get copied?
```
Comment (ButterscotchV) [inline server/src/main/java/io/github/axisangles/ktmath/Transform.kt:13]: Any explanation? What does this mean?

**Review by ButterscotchV — CHANGES_REQUESTED** [review]
Please note that while I have reviewed all AutoBone and PoseFrame changes, I have written the code myself, therefore I cannot provide a second opinion on them.

I also don't quite trust my review of `UnityArmature`, `HumanSkeleton`, or `LegTweaks`, they're huge classes, and not easy to compare on GitHub. I also couldn't properly review the settings for each `Tracker` made, as there are no argument names and I can't review them with my IDE at the moment.

File: server/src/main/java/dev/slimevr/tracking/trackers/TrackerPosition.java (hunk 7 above)
Comment (ImUrX) [reply server/src/main/java/dev/slimevr/tracking/trackers/TrackerPosition.java:135]: backwards compatibility of what version of slimevr?

File: server/src/main/java/io/github/axisangles/ktmath/Transform.kt (hunk 8 above)
Comment (ImUrX) [reply server/src/main/java/io/github/axisangles/ktmath/Transform.kt:13]: I'm copying a reference to the quaternion, and 2 vectors of the other transform. This is OK (because they are immutable) and at the same time makes you question that decision.

File: server/src/main/java/dev/slimevr/VRServer.java [hunk 9]
```diff
@@ -91,12 +90,22 @@ public VRServer(String configPath) {
...
+			"HMD",
+			"HMD",
+			TrackerPosition.HEAD,
+			true,
+			true,
+			false,
+			false,
+			false,
+			true
+		);
```
[the remark this answers is not in the document]
Comment (ImUrX) [reply server/src/main/java/dev/slimevr/VRServer.java:105]: the only thing we can do is add secondary constructors but that kinda ruins the purpose of it

[the remark this answers is not in the document]
Comment (ButterscotchV) [reply server/src/main/java/dev/slimevr/VRServer.java:105]: Then we'll add named arguments when we go to Kotlin uwu

File: server/src/main/java/com/jme3/math/Matrix4f.java [hunk 10]
```diff
@@ -1258,120 +1245,95 @@ public Vector4f multAcross(Vector4f vec, Vector4f store) {
...
-		float vx = vec.x, vy = vec.y, vz = vec.z;
-		store.x = m00 * vx + m01 * vy + m02 * vz + m03;
-		store.y = m10 * vx + m11 * vy + m12 * vz + m13;
-		store.z = m20 * vx + m21 * vy + m22 * vz + m23;
+	public float multProj(Vector3 vec) {
+		float vx = vec.getX(), vy = vec.getY(), vz = vec.getZ();
+		float x = m00 * vx + m01 * vy + m02 * vz + m03;
+		float y = m10 * vx + m11 * vy + m12 * vz + m13;
+		float z = m20 * vx + m21 * vy + m22 * vz + m23;
 		return m30 * vx + m31 * vy + m32 * vz + m33;
```
[the remark this answers is not in the document]
Comment (ButterscotchV) [reply server/src/main/java/com/jme3/math/Matrix4f.java:1288]: If it's unused then just remove it. :P
No use having a useless method eh?

File: server/src/main/java/dev/slimevr/tracking/trackers/TrackerPosition.java (hunk 7 above)
Comment (ButterscotchV) [reply server/src/main/java/dev/slimevr/tracking/trackers/TrackerPosition.java:135]: Very old, do you think it's no longer useful? I guess it's not really worthwhile to support such old versions. Feel free to resolve this if you agree.

Comment (Louka3000) [reply server/src/main/java/dev/slimevr/tracking/trackers/TrackerPosition.java:135]: There’s no reason to keep this much code for such old versions. Those specific trackers will just get unassigned is all. This code doesn’t even belong here in the first place; it should be in CurrentVRConfigConverter…

File: server/src/main/java/dev/slimevr/tracking/processor/BoneInfo.java (hunk 1 above)
Comment (ButterscotchV) [reply server/src/main/java/dev/slimevr/tracking/processor/BoneInfo.java:24]: If you think that this improves readability, then I'm fine with it. I just don't think needing euler angles for something so simple is necessarily beneficial otherwise.

File: server/src/main/java/com/jme3/math/Matrix4f.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (Louka3000) [reply server/src/main/java/com/jme3/math/Matrix4f.java:1288]: Do we remove all JME 3d math classes then? Vectors and Matrices that is.

File: server/src/main/java/io/github/axisangles/ktmath/Transform.kt (hunk 8 above)
Comment (ImUrX) [reply server/src/main/java/io/github/axisangles/ktmath/Transform.kt:13]: sigh just remove the fixme, it means transforms can be pointing to the same quaternion/vector3 object

File: server/src/main/java/dev/slimevr/tracking/processor/BoneInfo.java (hunk 1 above)
Comment (ImUrX) [reply server/src/main/java/dev/slimevr/tracking/processor/BoneInfo.java:24]: its converted to quaternion, there is nothing complex on this?

File: server/src/main/java/io/github/axisangles/ktmath/Transform.kt (hunk 8 above)
Comment (ButterscotchV) [reply server/src/main/java/io/github/axisangles/ktmath/Transform.kt:13]: Okay then that's fine, as long as it's immutable. You can leave a comment if you want, but I see a `FIXME`, so I'm curious what needs to be fixed, that's all.

File: server/src/main/java/com/jme3/math/FastMath.java [hunk 11]
```diff
@@ -454,29 +357,27 @@ public static Vector3f interpolateBezier(
...
-		Vector3f end = p2.clone();
 		if (endRange != 1) {
-			FastMath.interpolateCatmullRom(endRange, curveTension, p0, p1, p2, p3, end);
+			p2 = FastMath.interpolateCatmullRom(endRange, curveTension, p0, p1, p2, p3);
 		}
-		Vector3f middle = FastMath.interpolateCatmullRom(middleValue, curveTension, p0, p1, p2, p3);
-		float l = end.subtract(start).length();
-		float l1 = middle.subtract(start).length();
-		float l2 = end.subtract(middle).length();
+		Vector3 middle = FastMath.interpolateCatmullRom(middleValue, curveTension, p0, p1, p2, p3);
```
[the remark this answers is not in the document]
Comment (Louka3000) [reply server/src/main/java/com/jme3/math/FastMath.java:377]: Don't know how I messed that up ;-;

File: server/src/main/java/dev/slimevr/bridge/ProtobufBridge.java [hunk 12]
```diff
@@ -217,29 +223,29 @@ protected void trackerStatusReceived(TrackerStatus trackerStatus) {
...
-		for (ShareableTracker tracker : sharedTrackers) {
+		for (Tracker tracker : sharedTrackers) {
 			TrackerAdded.Builder builder = TrackerAdded
 				.newBuilder()
-				.setTrackerId(tracker.getTrackerId())
+				.setTrackerId(tracker.getId())
 				.setTrackerName(tracker.getName())
 				.setTrackerSerial(tracker.getName())
-				.setTrackerRole(tracker.getTrackerRole().id);
+				.setTrackerRole(tracker.getTrackerPosition().getTrackerRole().getId());
```
[the remark this answers is not in the document]
Comment (Louka3000) [reply server/src/main/java/dev/slimevr/bridge/ProtobufBridge.java:240]: Indeed, that's because ShareableTracker.java had the unique method `TrackerRole getTrackerRole()`. This new way of doing it is the propper way imo.

File: server/src/main/java/dev/slimevr/osc/VMCHandler.java [hunk 13]
```diff
@@ -376,34 +383,41 @@ public void update() {
...
-															.add(
-																outputUnityArmature
-																	.getHeadNodeOfBone(
-																		UnityBone.RIGHT_UPPER_LEG
-																	).worldTransform
-																		.getTranslation()
-															)).multLocal(0.5f)
-												))
+													)).times(0.5f)
+											))
```
[the remark this answers is not in the document]
Comment (Louka3000) [reply server/src/main/java/dev/slimevr/osc/VMCHandler.java:420]: ahah I kinda gave up on that code since it's bad anyway but sure

File: server/src/main/java/com/jme3/math/Matrix4f.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (ButterscotchV) [reply server/src/main/java/com/jme3/math/Matrix4f.java:1288]: Don't see why not as long as they aren't used, if we really think we *might* use something then we could make an exception for that, but this shows that it's obviously more of a burden to maintain with changes.

File: server/src/main/java/dev/slimevr/bridge/ProtobufBridge.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (ButterscotchV) [reply server/src/main/java/dev/slimevr/bridge/ProtobufBridge.java:238]: Got it, sounds good.

File: server/src/main/java/dev/slimevr/tracking/processor/BoneInfo.java (hunk 1 above)
Comment (ButterscotchV) [reply server/src/main/java/dev/slimevr/tracking/processor/BoneInfo.java:24]: It's not complex, just if it can be simpler then I'd take it as such. This is a non-issue.

File: server/src/main/java/dev/slimevr/posestreamer/BVHFileStream.java [hunk 14]
```diff
@@ -254,15 +253,15 @@ private void writeNodeHierarchyRotation(TransformNodeWrapper node, Quaternion in
...
 
 		// Adjust to local rotation
 		if (inverseRootRot != null) {
-			rotBuf = inverseRootRot.mult(rotBuf, rotBuf);
+			rot = inverseRootRot.times(rot);
 		}
 
 		// Roll (X), pitch (Y), yaw (Z) (intrinsic)
-		angleBuf = quatToXyzAngles(rotBuf.normalizeLocal(), angleBuf);
+		angleBuf = quatToXyzAngles(rot.unit(), angleBuf);
```
[the remark this answers is not in the document]
Comment (ButterscotchV) [reply server/src/main/java/dev/slimevr/posestreamer/BVHFileStream.java:264]:
```suggestion
		// TODO: `quatToXyzAngles` should no longer be needed, this should be
		// changed to use `Quaternion.toEulerAngles()` instead
		angleBuf = quatToXyzAngles(rot.unit(), angleBuf);
```

File: server/src/main/java/com/jme3/math/Matrix4f.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (Louka3000) [reply server/src/main/java/com/jme3/math/Matrix4f.java]: Yeah, and if we need them, we'll just make our own (well, Axis will)

File: server/src/main/java/dev/slimevr/bridge/ProtobufBridge.java [hunk 15]
```diff
@@ -81,11 +82,6 @@ public void dataRead() {
 			hadNewData = true;
 		}
 		if (hadNewData && hmdTracker != null) {
-			// If HMD was DISCONNECTED, need to rebuild the skeleton, as it
-			// ignores the HMD if it's DISCONNECTED in order to support using
-			// other trackers on head.
-			if (hmd.getStatus() == dev.slimevr.tracking.trackers.TrackerStatus.DISCONNECTED)
-				Main.getVrServer().updateSkeletonModel();
```
Comment (ButterscotchV) [inline server/src/main/java/dev/slimevr/bridge/ProtobufBridge.java:91]: Why was this removed? It doesn't seem too wrong but I'm curious of an explanation

File: server/src/main/java/dev/slimevr/osc/VRCOSCHandler.java [hunk 16]
```diff
@@ -236,14 +238,24 @@ public void update() {
...
+					// X quaternion represents a rotation from y to z
+					// Y quaternion represents a rotation from z to x
+					// When we negate the z direction, X and Y quaternion
+					// components must be negated.
+					EulerAngles unityAngles = new Quaternion(
+						rot.getW(),
+						-rot.getX(),
+						-rot.getY(),
+						rot.getZ()
+					).toEulerAngles(EulerOrder.YXZ);
```
Comment (ButterscotchV) [inline server/src/main/java/dev/slimevr/osc/VRCOSCHandler.java:254]: This... Seems really weird? Shouldn't the euler angles be the same no matter what as long as the order is correct?

**Review by ButterscotchV — CHANGES_REQUESTED** [review]

File: server/src/main/java/dev/slimevr/bridge/ProtobufBridge.java (hunk 15 above)
Comment (Louka3000) [reply server/src/main/java/dev/slimevr/bridge/ProtobufBridge.java:91]: It got moved to the tracker class directly, so it’s done for every tracker, not just HMD.

File: server/src/main/java/dev/slimevr/osc/VRCOSCHandler.java (hunk 16 above)
Comment (Louka3000) [reply server/src/main/java/dev/slimevr/osc/VRCOSCHandler.java:254]: No, because euler angle standards are all over the place even with the same order according to AxisAngle.

File: server/src/main/java/dev/slimevr/tracking/processor/skeleton/HumanSkeleton.java [hunk 17]
```diff
@@ -1072,20 +971,13 @@ else if (hasSpineTracker)
...
 		// R = InverseKnee * Ankle
 		// C = Quaternion(-R.x, 0, 0, R.w)
 		// Knee = Knee * R * C
 		// normalize(Knee)
-		rotBuf3.multLocal(ankle);
-		rotBuf4.set(-rotBuf3.getX(), 0, 0, rotBuf3.getW());
-		knee.multLocal(rotBuf3).multLocal(rotBuf4);
-		return knee.normalizeLocal();
+		Quaternion r = knee.inv().times(ankle);
+		Quaternion c = new Quaternion(r.getW(), -r.getX(), 0, 0);
```
Comment (ButterscotchV) [inline server/src/main/java/dev/slimevr/tracking/processor/skeleton/HumanSkeleton.java:979]: What's this line doing?

File: server/src/main/java/dev/slimevr/osc/VRCOSCHandler.java (hunk 16 above)
Comment (ButterscotchV) [reply server/src/main/java/dev/slimevr/osc/VRCOSCHandler.java:254]: That's stupid, I hate Euler angles. =w=
Make sure this is tested thoroughly, if it works and seems to be necessary, then that's fine with me.

File: server/src/main/java/dev/slimevr/bridge/ProtobufBridge.java (hunk 15 above)
Comment (ButterscotchV) [reply server/src/main/java/dev/slimevr/bridge/ProtobufBridge.java:91]: Ah, was thinking that, cool

File: server/src/main/java/dev/slimevr/osc/UnityArmature.kt [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+
+import com.jme3.math.FastMath
+import dev.slimevr.tracking.processor.TransformNode
+import io.github.axisangles.ktmath.EulerAngles
+import io.github.axisangles.ktmath.EulerOrder
+import io.github.axisangles.ktmath.Quaternion
+import io.github.axisangles.ktmath.Vector3
+
+private val LEFT_SHOULDER_OFFSET = EulerAngles(EulerOrder.YZX, 0f, 0f, FastMath.HALF_PI).toQuaternion()
+private val RIGHT_SHOULDER_OFFSET = EulerAngles(EulerOrder.YZX, 0f, 0f, -FastMath.HALF_PI).toQuaternion()
```
[the remark this answers is not in the document]
Comment (Louka3000) [reply server/src/main/java/dev/slimevr/osc/UnityArmature.kt:11]: They’re static yes. Why would we use a companion object? We don’t care about accessing them from the outside

File: server/src/main/java/dev/slimevr/osc/VRCOSCHandler.java (hunk 16 above)
Comment (Louka3000) [reply server/src/main/java/dev/slimevr/osc/VRCOSCHandler.java:254]: I went through it with AxisAngle and tested it yes ^^

File: server/src/main/java/dev/slimevr/osc/UnityArmature.kt [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+		rightShoulderTailNode.attachChild(rightElbowNode)
+		leftElbowNode.attachChild(leftWristNode)
+		rightElbowNode.attachChild(rightWristNode)
+		leftWristNode.attachChild(leftHandNode)
+		rightWristNode.attachChild(rightHandNode)
+	}
+
+	fun updateNodes() {
+		// Set the upper chest's rotation to the chest's
+		upperChestNode.localTransform.rotation = chestNode.localTransform.rotation
```
[the remark this answers is not in the document]
Comment (Louka3000) [reply server/src/main/java/dev/slimevr/osc/UnityArmature.kt:80]: We don’t have it ourselves but we need it for the translation

File: server/src/main/java/dev/slimevr/tracking/processor/skeleton/HumanSkeleton.java (hunk 17 above)
Comment (Louka3000) [reply server/src/main/java/dev/slimevr/tracking/processor/skeleton/HumanSkeleton.java:979]: Not sure what you mean? Same as before..

Comment (ButterscotchV) [reply server/src/main/java/dev/slimevr/tracking/processor/skeleton/HumanSkeleton.java:979]: Just curious about the math behind it, I know it's still the same, but I'm confused what its purpose really is

File: server/src/main/java/dev/slimevr/osc/UnityArmature.kt (hunk 19 above)
[the remark this answers is not in the document]
Comment (ButterscotchV) [reply server/src/main/java/dev/slimevr/osc/UnityArmature.kt:80]: So just basically `unity upper chest + chest = our chest`? If so, just resolve this comment.

File: server/src/main/java/dev/slimevr/tracking/processor/skeleton/HumanSkeleton.java (hunk 17 above)
Comment (Louka3000) [reply server/src/main/java/dev/slimevr/tracking/processor/skeleton/HumanSkeleton.java:979]: You’d have to ask AxisAngle about that one, I’m not able to describe *why* it works…

File: server/src/main/java/dev/slimevr/osc/UnityArmature.kt (hunk 18 above)
[the remark this answers is not in the document]
Comment (ButterscotchV) [reply server/src/main/java/dev/slimevr/osc/UnityArmature.kt:11]: > We don’t care about accessing them from the outside

Exactly, defining them within the namespace puts the variables throughout the *entire* namespace, whereas a companion object will keep it only defined within the class. Both are accessible from the outside, but the scope is different.

File: server/src/main/java/dev/slimevr/tracking/processor/skeleton/HumanSkeleton.java (hunk 17 above)
Comment (ButterscotchV) [reply server/src/main/java/dev/slimevr/tracking/processor/skeleton/HumanSkeleton.java:979]: I'll resolve this and ask more generall on the Discord then, this is not relevant to this PR.

File: server/src/main/java/dev/slimevr/osc/UnityArmature.kt (hunk 18 above)
[the remark this answers is not in the document]
Comment (Louka3000) [reply server/src/main/java/dev/slimevr/osc/UnityArmature.kt]: fixed

**Comment by Louka3000** [comment]
Tracker resets are broken.
ZRock is able to demonstrate and replicate the problem, but I haven't gotten much from debugging from him (reset quaternions seem fine)
Butter is investigating, and so will I when I get trackers and am able to replicate.

File: server/src/main/java/io/eiren/math/Vector3d.java [hunk 20]
```diff
@@ -1,6 +1,7 @@
 package io.eiren.math;
 
-import com.jme3.math.Vector3f;
+
+import io.github.axisangles.ktmath.Vector3;
 
 
 public class Vector3d implements Cloneable {
```
Comment (Eirenliel) [inline server/src/main/java/io/eiren/math/Vector3d.java:7]: Is this used? We can probably remove the class.

File: server/src/main/java/io/github/axisangles/ktmath/Transform.kt (hunk 8 above)
Comment (Eirenliel) [inline server/src/main/java/io/github/axisangles/ktmath/Transform.kt:13]:
```suggestion
		// Quaternions and vectors are immutable, so we're copying links
```

File: server/src/main/java/dev/slimevr/tracking/processor/skeleton/HumanSkeleton.java [hunk 21]
```diff
@@ -1072,20 +971,13 @@ else if (hasSpineTracker)
...
 	 */
 	private Quaternion extendedKneeYawRoll(Quaternion knee, Quaternion ankle) {
-		// Clone the knee since we're modifying it and returning it
-		knee = knee.clone();
-
-		// Get the inverse rotation of the knee
-		rotBuf3.set(knee).inverseLocal();
-
 		// R = InverseKnee * Ankle
 		// C = Quaternion(-R.x, 0, 0, R.w)
```
Comment (Eirenliel) [inline server/src/main/java/dev/slimevr/tracking/processor/skeleton/HumanSkeleton.java:975]:
```suggestion
		// C = Quaternion(R.w, -R.x, 0, 0)
```

File: server/src/main/java/dev/slimevr/tracking/processor/skeleton/HumanSkeleton.java [hunk 22]
```diff
@@ -1102,117 +994,91 @@ private Quaternion extendedPelvisYawRoll(
...
-			rightKnee.negateLocal();
+		if (kneeRot.dot(rightKnee) < 0.0f) {
+			rightKnee = rightKnee.unaryMinus();
 		}
 
-		// Get the inverse rotation of the hip.
-		rotBuf1.set(hip).inverseLocal();
-
 		// R = InverseHip * (LeftLeft + RightLeg)
 		// C = Quaternion(-R.x, 0, 0, R.w)
```
Comment (Eirenliel) [inline server/src/main/java/dev/slimevr/tracking/processor/skeleton/HumanSkeleton.java:1008]:
```suggestion
		// C = Quaternion(R.w, -R.x, 0, 0)
```

File: server/src/main/java/io/eiren/math/Vector3d.java (hunk 20 above)
Comment (Louka3000) [reply server/src/main/java/io/eiren/math/Vector3d.java:7]: Done

File: server/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweakBuffer.java [hunk 23]
```diff
@@ -765,18 +729,18 @@ private float[] getPressurePrediction() {
 
 		// set the pressure to the force on each foot times the force to
 		// pressure scalar
-		leftFootPressure = leftFootForce.mult(FORCE_VECTOR_TO_PRESSURE).length();
-		rightFootPressure = rightFootForce.mult(FORCE_VECTOR_TO_PRESSURE).length();
+		leftFootPressure = leftFootForce.cross(FORCE_VECTOR_TO_PRESSURE).len();
+		rightFootPressure = rightFootForce.cross(FORCE_VECTOR_TO_PRESSURE).len();
```
[the remark this answers is not in the document]
Comment (Louka3000) [reply server/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweakBuffer.java:733]: Cross is for rotation vectors (same as old mult) whereas times just multiplies each components. Which one would be the intended one here?

File: server/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweaks.java [hunk 24]
```diff
@@ -475,31 +464,54 @@ public void tweakLegs() {
...
+			);
+			rightKneePosition = new Vector3(
+				rightKneePosition.getX(),
+				bufferHead.getLeftKneePosition().getY(),
+				rightKneePosition.getZ()
+			);
 		}
 
 		// calculate the correction for the knees
-		if (kneesActive && initialized)
```
[the remark this answers is not in the document]
Comment (Louka3000) [reply server/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweaks.java:488]: Yes, the computed skeleton trackers are never null.

File: server/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweakBuffer.java (hunk 23 above)
[the remark this answers is not in the document]
Comment (Stermere) [reply server/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweakBuffer.java:733]: Times is the intended there, it's to scale the vector so that the y component shines through more in the pressure

File: server/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweaks.java (hunk 24 above)
[the remark this answers is not in the document]
Comment (Stermere) [reply server/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweaks.java:488]: great just checking

File: server/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweakBuffer.java (hunk 23 above)
[the remark this answers is not in the document]
Comment (Louka3000) [reply server/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweakBuffer.java:733]: Oh wait I was wrong, that's not how it works, lemme ask Uriel

[the remark this answers is not in the document]
Comment (Louka3000) [reply server/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweakBuffer.java:733]: Ok it's hadamard lol, fixed it.

**Review by ButterscotchV — APPROVED** [review]

**Review by ButterscotchV — APPROVED** [review]

File: server/src/main/java/dev/slimevr/tracking/trackers/udp/TrackersUDPServer.java [hunk 25]
```diff
@@ -331,16 +340,14 @@ public void run() {
 							} else {
 								conn.timedOut = false;
 								for (Tracker value : conn.getTrackers().values()) {
-									IMUTracker tracker = (IMUTracker) value;
-									if (tracker.getStatus() == TrackerStatus.DISCONNECTED)
-										tracker.setStatus(TrackerStatus.OK);
+									value.setStatus(TrackerStatus.OK);
```
Comment (Eirenliel) [inline server/src/main/java/dev/slimevr/tracking/trackers/udp/TrackersUDPServer.java:343]: this will reset error state, should only set to OK if it was DISCONNECTED previously

File: server/src/main/java/dev/slimevr/tracking/trackers/udp/TrackersUDPServer.java [hunk 26]
```diff
@@ -317,12 +325,13 @@ public void run() {
...
+							socket
+								.send(
+									new DatagramPacket(rcvBuffer, bb.position(), conn.getAddress())
+								);
 							if (conn.lastPacket + 1000 < System.currentTimeMillis()) {
 								for (Tracker value : conn.getTrackers().values()) {
-									IMUTracker tracker = (IMUTracker) value;
-									if (tracker.getStatus() == TrackerStatus.OK)
-										tracker.setStatus(TrackerStatus.DISCONNECTED);
+									value.setStatus(TrackerStatus.DISCONNECTED);
```
Comment (Eirenliel) [inline server/src/main/java/dev/slimevr/tracking/trackers/udp/TrackersUDPServer.java:334]: this will override error state, that's why there was a check before

**Review by Eirenliel — CHANGES_REQUESTED** [review]

**Review by Eirenliel — APPROVED** [review]

File: server/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweaks.java (hunk 24 above)
[the remark this answers is not in the document]
Comment (Louka3000) [reply server/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweaks.java:488]: It was in fact, not a good idea to remove it. _However, its assignment did have to change as computed trackers being non-null would've always made it active, giving the same behaviour as when I removed it :p_
